### PR TITLE
add ONNX support for BLOOM

### PR DESCRIPTION
**Pull request by nmntz**
# What does this PR do?

add ONNX support for BLOOM

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@michaelbenayoun

### CODE REVIEW

**Comment by younesbelkada** [comment]
As you told me offline that the slow tests were passing (under torch1.11.0), looks good to me! Thanks for working on that 🔥

File: src/transformers/models/bloom/configuration_bloom.py
Comment (lewtun) [inline src/transformers/models/bloom/configuration_bloom.py:164]: Nice!

File: src/transformers/models/bloom/configuration_bloom.py
Comment (lewtun) [inline src/transformers/models/bloom/configuration_bloom.py:175]: Hehe @michaelbenayoun we should fix this sometime :)

File: tests/onnx/test_onnx_v2.py
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:207]: Have you checked that the slow tests pass for this checkpoint? You can run:

```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "bloom"
```

**Review by lewtun — APPROVED** [review]
Thanks for adding the Bloom config @NouamaneTazi - a very clean PR 💮 !

If the slow tests pass, this PR looks good to me. Let's wait for approval from @LysandreJik or @sgugger before merging this

Edit: I see the CI is now failing for an unrelated issue. I've re-run it, but if it comes up red again, I suggest rebasing on `main` and pushing again

File: src/transformers/models/bloom/configuration_bloom.py
Comment (michaelbenayoun) [inline src/transformers/models/bloom/configuration_bloom.py:19]: I would use strings for type annotations and something like:
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
   from transformers import ...
```
That way you don't import classes that are not actually being used.

File: src/transformers/models/bloom/modeling_bloom.py
Comment (michaelbenayoun) [inline src/transformers/models/bloom/modeling_bloom.py:298]: Dont know if this is relevant but the original implementation outputs a tensor of rank 2, and your change outputs a tensor of rank 3. Should not be a big deal since we do reshape it afterwards but just wanted to point this out.

**Review by michaelbenayoun — APPROVED** [review]
Great, thanks for handling this!

**Comment by younesbelkada** [comment]
@sgugger I do not think this will hurt performances in terms of logits since slow tests are passing, but might hurt indeed the inference time performance for large and/or batched sequences.. We need to benchmark that though to be sure

File: tests/onnx/test_onnx_v2.py
Comment (NouamaneTazi) [reply tests/onnx/test_onnx_v2.py:207]: Thank you for reminding me. All tests are passing now 🙂

File: src/transformers/models/bloom/modeling_bloom.py
Comment (michaelbenayoun) [inline src/transformers/models/bloom/modeling_bloom.py:84]: This looks good!

File: src/transformers/models/bloom/modeling_bloom.py
Comment (michaelbenayoun) [reply src/transformers/models/bloom/modeling_bloom.py:298]: Do we keep it like that?

File: src/transformers/models/bloom/configuration_bloom.py
Comment (michaelbenayoun) [inline src/transformers/models/bloom/configuration_bloom.py:225]: I think you mention in one commit that the past shape is not the same as the one in `gpt-2`, are the dynamic axes left unchanged though? Asking because otherwise it can break you `inputs` property.

**Review by michaelbenayoun — APPROVED** [review]
LGTM

**Comment by michaelbenayoun** [comment]
Also make sure all the tests pass before merging.

File: src/transformers/models/bloom/configuration_bloom.py
Comment (NouamaneTazi) [reply src/transformers/models/bloom/configuration_bloom.py:225]: Yes, the aim of that commit was to fix the `past_key_values` length.
I did make sure that the inputs work properly now, and the tests are passing

**Comment by NouamaneTazi** [comment]
All tests for `tests/onnx/test_onnx_v2.py -k "bloom"` and `tests/models/bloom` are passing.
Here are the ones that are skipped (which is fine according to @younesbelkada)
```
================================================================================= short test summary info =================================================================================
SKIPPED [1] tests/test_modeling_common.py:2006: test is PT+FLAX test
SKIPPED [1] tests/test_modeling_common.py:1934: test is PT+FLAX test
SKIPPED [1] tests/test_modeling_common.py:1758: test is PT+TF test
SKIPPED [1] tests/test_tokenization_common.py:1960: This test is only for slow tokenizers
SKIPPED [1] tests/test_tokenization_common.py:2189: test is PT+TF test
================================================================= 159 passed, 5 skipped, 35 warnings in 449.50s (0:07:29)
```

**Comment by sgugger** [comment]
There is a difference between a copy in BLOOM and the original in GPT-2 which is why the CI is failing. Make sure to run `make fic-copies` or remove the `Copied from`.

File: src/transformers/models/bloom/configuration_bloom.py
Comment (sgugger) [inline src/transformers/models/bloom/configuration_bloom.py:168]: The comment can be removed entirely then, there is no point leaving it
```suggestion
```

File: src/transformers/models/bloom/modeling_bloom.py
Comment (NouamaneTazi) [reply src/transformers/models/bloom/modeling_bloom.py]: Fixed :) Thank you for the notice 🤗

**Review by sgugger — APPROVED** [review]
Thanks a lot!

**Pull request merged**
